### ISSUE TITLE
[Parity 14/16] torch.nn primitives gap - parametrizations, pruning, losses, utilities

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#223]
Part of epic #209.

## Summary

This issue covers `torch.nn` primitives that live at the tensor-library layer: parametrizations, pruning APIs, and the long-tail of missing losses and utilities. Core layer modules (Linear / Conv / Transformer / etc.) live upstream in AiDotNet core — this issue is about the *functional* surface and utilities that must exist in the tensor library for the upstream layers to use.

### Parametrizations (`torch.nn.utils.parametrize`)
- `register_parametrization(module, tensor_name, parametrization)` — wrap a parameter with a reparam (e.g. `exp` for positive params)
- Built-in parametrizations:
  - `weight_norm` (factored into direction + magnitude)
  - `spectral_norm` (power-iteration spectral normalization)
  - `orthogonal` (Householder / Cayley / matrix-exp orthogonalization)

### Pruning (`torch.nn.utils.prune`)
- `prune.l1_unstructured`, `prune.random_unstructured`
- `prune.ln_structured`, `prune.random_structured`
- `prune.global_unstructured` (top-K across multiple modules)
- `prune.custom_from_mask`
- Pruning lifecycle: forward-pre-hook that zeros masked weights
- `prune.remove` — consolidate mask into the weight

### Missing losses (compared to `torch.nn.functional`)
- `ctc_loss` (with zero_infinity, reduction)
- `huber_loss` (delta)
- `smooth_l1_loss` (beta)
- `poisson_nll_loss` (log_input, full, eps, reduction)
- `gaussian_nll_loss` (full, eps)
- `multi_label_margin_loss`, `multi_label_soft_margin_loss`
- `multi_margin_loss`
- `triplet_margin_loss`, `triplet_margin_with_distance_loss`
- `margin_ranking_loss`
- `cosine_embedding_loss`
- `hinge_embedding_loss`
- `kl_div` (already probably covered, but with `log_target`)

### Utility functions
- `F.normalize(input, p, dim, eps)` (p-norm normalize)
- `F.embedding`, `F.embedding_bag` (with max_norm, norm_type, scale_grad_by_freq, sparse gradients)
- `F.one_hot`
- `F.cosine_similarity`
- `F.pairwise_distance`
- `F.pdist`, `F.cdist`

### Activation completeness
- Mish, Hardshrink, Softshrink, Tanhshrink, Threshold, PReLU, RReLU (randomized ReLU)
- Newer: `GELU(approximate="tanh")`, `QuickGELU`, `SwiGLU`, `GEGLU`, `ReGLU`

## Current state in the repo

- `ActivationRegistry` exists (per CLAUDE.md).
- No parametrization framework.
- No pruning API.
- Many losses — surface unconfirmed; treat as missing pending audit.
- Missing: `normalize`, `embedding_bag`, `pdist`/`cdist`, PReLU/RReLU.

## How we beat PyTorch

1. **Parametrizations that compose with compile**: `spectral_norm` via power iteration as a `LazyNode` so it fuses with the subsequent matmul. PyTorch's `spectral_norm` is a forward hook that adds a materialization step.
2. **Orthogonal parametrization with matrix-exp path** (needs linalg issue's `matrix_exp`) — the mathematically cleanest option. PyTorch supports it but in an opt-in branch.
3. **Pruning API integrated with our `SymbolicBackwardGraph`**: a pruned parameter with a compile-time-known mask lets the backward skip the masked branch. PyTorch's pruning is a runtime multiply.
4. **Global unstructured pruning by magnitude via radix-sort** (needs tensor-ops issue's sort). Fast cross-module top-K.
5. **SIMD CTC loss**: PyTorch's CPU CTC is painful. Our AVX-512 scan-based forward + backward targets 3× on speech workloads.
6. **Embedding bag with sub-byte storage**: 1-bit / 2-bit / 4-bit embeddings natively supported (no dequant-before-lookup). Crushing for RecSys.
7. **Fused `normalize` + `matmul` + `bias` + `activation`** via `FusedOperations.cs` — the common contrastive-learning path becomes one kernel.
8. **`pdist`/`cdist` using our GEMM + `sqrt` pipeline**: `||x_i - x_j||² = ||x_i||² + ||x_j||² - 2 x_i·x_j` → one GEMM + two norms. Target 2× PyTorch.
9. **All newer GLU variants (SwiGLU, GEGLU, ReGLU, QuickGELU)** as first-class activations — PyTorch users compose them from primitives.

### Parametrizations
- [ ] `Parametrize(tensor, reparam)` framework
- [ ] `WeightNorm` parametrization
- [ ] `SpectralNorm` parametrization
- [ ] `Orthogonal` parametrization (Householder, Cayley, matrix-exp)
- [ ] `Parametrize.Remove(...)` consolidation

### Pruning
- [ ] `Prune.L1Unstructured(amount)`, `RandomUnstructured(amount)`
- [ ] `Prune.LnStructured(amount, n, dim)`, `RandomStructured(amount, dim)`
- [ ] `Prune.GlobalUnstructured(modules, amount, method)`
- [ ] `Prune.CustomFromMask(tensor, mask)`
- [ ] `Prune.Remove(module, name)` consolidation
- [ ] Pre-forward hook or compile-mode mask fusion

### Losses
- [ ] `CtcLoss` (CPU AVX-512 + CUDA via cuDNN from GPU-primitives issue)
- [ ] `HuberLoss`, `SmoothL1Loss`
- [ ] `PoissonNllLoss`, `GaussianNllLoss`
- [ ] `MultiLabelMarginLoss`, `MultiLabelSoftMarginLoss`, `MultiMarginLoss`
- [ ] `TripletMarginLoss`, `TripletMarginWithDistanceLoss`
- [ ] `MarginRankingLoss`
- [ ] `CosineEmbeddingLoss`
- [ ] `HingeEmbeddingLoss`
- [ ] `KLDiv` with `log_target` variant

### Utility functions
- [ ] `Normalize(p, dim, eps)`
- [ ] `Embedding` (with max_norm, norm_type, scale_grad_by_freq, sparse grads)
- [ ] `EmbeddingBag` (with sub-byte storage bonus)
- [ ] `OneHot`
- [ ] `CosineSimilarity`
- [ ] `PairwiseDistance`
- [ ] `PDist`, `CDist`

### Activations
- [ ] Audit existing `ActivationRegistry`
- [ ] Mish, Hardshrink, Softshrink, Tanhshrink, Threshold, PReLU, RReLU
- [ ] GELU tanh approximation, QuickGELU
- [ ] SwiGLU, GEGLU, ReGLU as first-class ops

## Acceptance criteria

- Parametrizations pass correctness + gradient tests.
- Pruning: correctness, lifecycle tests, mask-remove finalization.
- Every loss: reference parity (within tolerance), gradcheck.
- Utility functions: parity + gradcheck + tensor-parity tests.
- Benchmarks: CTC on CPU ≥ 3× PyTorch; pdist/cdist ≥ 2× PyTorch.

## Implementation notes

- New module `src/AiDotNet.Tensors/NN/` (or align with existing naming in `Engines/` if it fits better).
- Losses under `NN/Losses/`, utilities under `NN/Functional/`, parametrizations under `NN/Parametrizations/`, pruning under `NN/Pruning/`.
- All ops registered with `OpRegistry` + backward in `DifferentiableOps`.
- Sub-byte EmbeddingBag reuses `PackedInt1/2/3/4.cs`.

**Comment by ooples** [comment]
# Implementation plan — detailed

## 1. Current state

`ActivationRegistry` exists. No parametrization framework, no pruning API, many losses absent, `normalize`/`embedding_bag`/`pdist`/`cdist` missing.

## 2. Scope — full inventory

### Parametrizations
```
Parametrize(tensor, reparam) framework
  - register / remove / is_parametrized
WeightNorm parametrization
SpectralNorm (power iteration, n_power_iterations, eps, dim)
Orthogonal (Householder, Cayley, matrix_exp — needs #211)
```

### Pruning
```
Prune.L1Unstructured(amount)
Prune.RandomUnstructured(amount)
Prune.LnStructured(amount, n, dim)
Prune.RandomStructured(amount, dim)
Prune.GlobalUnstructured(modules, amount, method)
Prune.CustomFromMask(tensor, mask)
Prune.Remove(module, name) — consolidate mask into weight
```

### Losses
```
CtcLoss (zero_infinity, reduction, blank)
HuberLoss (delta, reduction)
SmoothL1Loss (beta, reduction)
PoissonNllLoss (log_input, full, eps, reduction)
GaussianNllLoss (full, eps, reduction)
MultiLabelMarginLoss (reduction)
MultiLabelSoftMarginLoss
MultiMarginLoss (p, margin, weight)
TripletMarginLoss (margin, p, swap)
TripletMarginWithDistanceLoss (distance_fn, margin, swap)
MarginRankingLoss (margin)
CosineEmbeddingLoss (margin)
HingeEmbeddingLoss (margin)
KLDiv with log_target variant
```

### Utility functions
```
Normalize(input, p, dim, eps)
Embedding(weight, input, padding_idx, max_norm, norm_type, scale_grad_by_freq, sparse)
EmbeddingBag(weight, input, offsets, max_norm, norm_type, scale_grad_by_freq, mode,
             sparse, per_sample_weights, include_last_offset)
OneHot(input, num_classes)
CosineSimilarity(x1, x2, dim, eps)
PairwiseDistance(x1, x2, p, eps, keepdim)
PDist(input, p)
CDist(x1, x2, p, compute_mode)
```

### Activation completeness
```
Mish, Hardshrink, Softshrink, Tanhshrink, Threshold, PReLU, RReLU
GELU(approximate="tanh"), QuickGELU
SwiGLU, GEGLU, ReGLU — first-class
```

**Total: ~45 ops.**

## 3. Win vectors

1. **Parametrizations compose with compile**: `spectral_norm` via power iteration as `LazyNode` → fuses with subsequent matmul. PyTorch's is forward hook with materialization.
2. **Orthogonal via `matrix_exp`** (needs #211) — mathematically cleanest.
3. **Pruning integrated with `SymbolicBackwardGraph`**: pruned param with compile-known mask → backward skips masked branch. PyTorch is runtime multiply.
4. **Global unstructured pruning via radix-sort** (needs #210).
5. **SIMD CTC loss**: AVX-512 scan-based forward + backward. Target 3×.
6. **Embedding bag with sub-byte storage** (1/2/4-bit) — no dequant-before-lookup. RecSys kill-shot.
7. **Fused `normalize + matmul + bias + activation`** via `FusedOperations.cs`.
8. **`pdist`/`cdist` via GEMM + sqrt**: ‖x_i−x_j‖² = ‖x_i‖² + ‖x_j‖² − 2 x_i·x_j. Target 2×.
9. **All newer GLU variants first-class** (SwiGLU/GEGLU/ReGLU/QuickGELU).

## 4. Per-axis beat targets

| Axis | Target |
|---|---|
| CPU CTC loss | ≥3× PyTorch CPU |
| `pdist`/`cdist` | ≥2× PyTorch |
| `embedding_bag` with int4 storage | works; feature unique |
| `spectral_norm` in compiled plan | fuses with matmul (PyTorch doesn't) |
| `orthogonal` via matrix_exp | works with gradcheck |
| Global unstructured pruning | via radix-sort; ≥2× PyTorch sort path |
| All 6 GPU backends | CPU + CUDA required; others best-effort |
| Autograd | gradcheck for every differentiable op |

## 5. File inventory

**New (~70):**
- `src/AiDotNet.Tensors/NN/` root
- `NN/Parametrizations/` — Parametrize, WeightNorm, SpectralNorm, Orthogonal (~6)
- `NN/Pruning/` — Prune.*, PruningHook, MaskStore (~10)
- `NN/Losses/` — CtcLoss, HuberLoss, SmoothL1, PoissonNll, GaussianNll, MultiLabel*, MultiMargin, TripletMargin*, MarginRanking, CosineEmbedding, HingeEmbedding (~15)
- `NN/Functional/` — Normalize, Embedding, EmbeddingBag (sub-byte), OneHot, CosineSimilarity, PairwiseDistance, PDist, CDist (~10)
- `NN/Activations/` — Mish, Hardshrink, Softshrink, Tanhshrink, Threshold, PReLU, RReLU, QuickGELU, SwiGLU, GEGLU, ReGLU + tanh-approx GELU (~12)
- Per-backend GPU kernels (~20)

**Modified (~8):** `ActivationRegistry.cs`, `OpRegistry.cs`, `DifferentiableOps.cs`, `BackwardFunctions.cs`, `FusedOperations.cs`, `PackedInt*.cs` (EmbeddingBag sub-byte).

**Tests (~30):** parametrization correctness + gradient, pruning lifecycle + mask-remove, per-loss parity + gradcheck, utility parity + gradcheck, activation parity.

**Benchmarks (~5):** `CtcLossBenchmarks`, `PdistCdistBenchmarks`, `EmbeddingBagBenchmarks`, `SpectralNormCompiledBenchmarks`, `GLUVariantsBenchmarks`.

## 6. Acceptance gate

1. All parametrizations pass correctness + gradient tests; orthogonal via matrix_exp works.
2. Pruning lifecycle + consolidation work; global unstructured via radix sort.
3. Every loss: reference parity + gradcheck.
4. Utility functions: parity + gradcheck.
5. CTC on CPU ≥3× PyTorch.
6. pdist/cdist ≥2× PyTorch.
7. EmbeddingBag with int4 storage demonstrated.
8. All 6 GPU backends ship CPU-equivalent or skip-with-reason.

## 7. Timeline — ~5 weeks

W1 Parametrization framework + WeightNorm + SpectralNorm + Orthogonal.
W2 Pruning framework + all methods.
W3 Losses (11 ops).
W4 Utility functions + activation completeness + GLU variants.
W5 Benchmarks + GPU kernels + PR.

## 8. Dependencies

- **Blocks**: nothing hard.
- **Blocked by**: #210 (radix sort for global pruning; Unique for pruning methods), #211 (matrix_exp for Orthogonal), #214 (forward-mode jvp for parametrizations).

## 9. Risks

- **CTC backward** numerically tricky at log-domain transitions; reference CuDNN impl.
- **EmbeddingBag `max_norm`** has side effects — careful with tape semantics.
- **Pruning + FSDP** interaction needs test coverage once both land.

---

*Part of epic #209. Plan ratified 2026-04-19.*
